grakn-engine/src/main/java/ai/grakn/engine/controller/GraqlController.java

```diff
@@ -95,39 +103,68 @@ public GraqlController(EngineGraknGraphFactory factory, Service spark) {
             @ApiImplicitParam(name = KEYSPACE,    value = "Name of graph to use", required = true, dataType = "string", paramType = "query"),
             @ApiImplicitParam(name = QUERY,       value = "Match query to execute", required = true, dataType = "string", paramType = "query"),
             @ApiImplicitParam(name = INFER,       value = "Should reasoner with the current query.", required = true, dataType = "boolean", paramType = "query"),
-            @ApiImplicitParam(name = MATERIALISE, value = "Should reasoner materialise results with the current query.", required = true, dataType = "boolean", paramType = "query")
     })
-    private Json executeGraql(Request request, Response response){
+    private Json executeGraqlGET(Request request, Response response){
         String keyspace = mandatoryQueryParameter(request, KEYSPACE);
         String queryString = mandatoryQueryParameter(request, QUERY);
         boolean infer = parseBoolean(mandatoryQueryParameter(request, INFER));
         boolean materialise = parseBoolean(mandatoryQueryParameter(request, MATERIALISE));
-        Json responseBody;
+        String acceptType = getAcceptType(request);
 
         try(GraknGraph graph = factory.getGraph(keyspace, WRITE)){
             Query<?> query = graph.graql().materialise(materialise).infer(infer).parse(queryString);
 
-            if(!readOnly(query)){
+            if(!query.isReadOnly()){
                 throw new GraknEngineServerException(405, "Only \"read-only\" queries are allowed.");
             }
-            //Do not return directly inside the try block otherwise open transactions will not autocommit and close properly.
-            responseBody = respond(request, query, response);
+
+            if(!validContentType(acceptType, query)){
+                throw new GraknEngineServerException(406, INVALID_CONTENT_TYPE, query.getClass().getName(), acceptType);
+            }
+
+            Json responseBody = executeReadQuery(request, query, acceptType);
+            return respond(response, query, acceptType, responseBody);
         }
+    }
 
-        return responseBody;
+    @POST
+    @Path("/")
+    @ApiOperation(
+            value = "Executes graql insert query on the server and returns the IDs of the inserted concepts.")
+    @ApiImplicitParams({
+            @ApiImplicitParam(name = KEYSPACE,    value = "Name of graph to use", required = true, dataType = "string", paramType = "query"),
+            @ApiImplicitParam(name = QUERY,       value = "Insert query to execute", required = true, dataType = "string", paramType = "body"),
+    })
+    private Json executeGraqlPOST(Request request, Response response){
+        String queryString = mandatoryBody(request);
+        String keyspace = mandatoryQueryParameter(request, KEYSPACE);
+
+        try(GraknGraph graph = factory.getGraph(keyspace, WRITE)){
+            Query<?> query = graph.graql().materialise(false).infer(false).parse(queryString);
```

Comment:
Validation exceptions will be caught by an error handlers, yes, but they won't return their own code. Ill make that change now.